src/Temporalio/Bridge/ByteArrayRef.cs

```diff
@@ -0,0 +1,112 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Google.Protobuf;
+
+namespace Temporalio.Bridge
+{
+    internal class ByteArrayRef
+    {
+        public static readonly ByteArrayRef Empty = new(new byte[0]);
+        internal static readonly UTF8Encoding StrictUTF8 = new(false, true);
+
+        public static ByteArrayRef FromUTF8(string s)
+        {
+            if (s.Length == 0)
+            {
+                return Empty;
+            }
+            return new ByteArrayRef(StrictUTF8.GetBytes(s));
+        }
+
+        public static ByteArrayRef FromProto(IMessage p)
+        {
+            return new ByteArrayRef(p.ToByteArray());
+        }
+
+        public static ByteArrayRef FromMetadata(IEnumerable<KeyValuePair<string, string>> metadata)
+        {
+            var stream = new MemoryStream();
+            using (var writer = new StreamWriter(stream, StrictUTF8))
+            {
+                foreach (var pair in metadata)
+                {
+                    // If either have a newline, we error since it would make an invalid set
+                    if (pair.Key.IndexOf('\n') >= 0 || pair.Value.IndexOf('\n') >= 0)
+                    {
+                        throw new ArgumentException("Metadata keys/values cannot have newlines");
```

Comment:
Because that limitation already exists for gRPC metadata and HTTP headers (e.g. https://github.com/grpc/grpc-go/issues/1888 and other places) and I needed some reasonable delimiter.